src/pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/goharbor/harbor/src/pkg/scheduler/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	jobParamCallbackFunc       = "callback_func"
+	jobParamCallbackFuncParams = "params"
+)
+
+var (
+	registry = make(map[string]CallbackFunc)
+)
+
+// CallbackFunc defines the function that the scheduler calls when triggered
+type CallbackFunc func(interface{}) error
+
+// Scheduler provides the capability to run a periodic task, a callback function
+// needs to be registered before using the scheduler
+type Scheduler interface {
+	Schedule(cron string, callbackFuncName string, params interface{}) (int64, error)
+	UnSchedule(id int64) error
+}
+
+// Register the callback function with name, and the function will be called
+// by the scheduler when the scheduler is triggered
+func Register(name string, callbackFunc CallbackFunc) error {
+	if len(name) == 0 {
+		return errors.New("empty name")
+	}
+	if callbackFunc == nil {
+		return errors.New("callback function is nil")
+	}
+
+	_, exist := registry[name]
+	if exist {
+		return fmt.Errorf("callback function %s already exists", name)
+	}
+	registry[name] = callbackFunc
+
+	return nil
+}
+
+// GetCallbackFunc returns the registered callback function specified by the name
+func GetCallbackFunc(name string) (CallbackFunc, error) {
+	f, exist := registry[name]
+	if !exist {
+		return nil, fmt.Errorf("callback function %s not found", name)
+	}
+	return f, nil
+}
+
+func callbackFuncExist(name string) bool {
+	_, exist := registry[name]
+	return exist
+}
+
+// New returns an instance of the default scheduler
+func New() Scheduler {
+	internalCoreURL := config.InternalCoreURL()
+	coreSecret := config.CoreSecret()
+	internalJobserviceURL := config.InternalJobServiceURL()
+	jobserviceClient := cjob.NewDefaultClient(internalJobserviceURL, coreSecret)
+	return &scheduler{
+		internalCoreURL:  internalCoreURL,
+		jobserviceClient: jobserviceClient,
+		manager:          NewManager(),
+	}
+}
+
+type scheduler struct {
+	sync.RWMutex
+	internalCoreURL  string
+	manager          Manager
+	jobserviceClient cjob.Client
+}
+
+func (s *scheduler) Schedule(cron string, callbackFuncName string, params interface{}) (int64, error) {
+	if !callbackFuncExist(callbackFuncName) {
+		return 0, fmt.Errorf("callback function %s not found", callbackFuncName)
+	}
+
+	// submit scheduler job to Jobservice
+	statusHookURL := fmt.Sprintf("%s/service/notifications/scheduler", s.internalCoreURL)
+	jd := &models.JobData{
+		Name: JobNameScheduler,
+		Parameters: map[string]interface{}{
+			jobParamCallbackFunc:       callbackFuncName,
+			jobParamCallbackFuncParams: params,
+		},
+		Metadata: &models.JobMetadata{
+			JobKind: job.JobKindPeriodic,
+			Cron:    cron,
+		},
+		StatusHook: statusHookURL,
+	}
+	jobID, err := s.jobserviceClient.SubmitJob(jd)
+	if err != nil {
+		return 0, err
+	}
+	log.Debugf("the scheduler job submitted to Jobservice, job ID: %s", jobID)
+
+	// create schedule record
+	now := time.Now()
```

Comment:
Can we create record first and then submit the job? If we failed to submit the job, then mark the status to error.